### PR TITLE
grpc: Use lru cache for grpc raw async client

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
Commit Message: As we are migrating to cache raw async grpc client, it is necessary to evict unused client to prevent the size of the cache to grow infinitely.
Additional Description:
Risk Level: low
Testing: Some test to verify the behavior.
Docs Changes:NA
Release Notes:NA
Platform Specific Features:NA

### CODE REVIEW

**Comment by jmarantz** [comment]
I don't understand the rationale.

How do you know what the right number of entries to keep is?

It seems not at all difficult to bound the churn of opening/closing new connections by keeping them open for X seconds. That's pretty common practice in networking (e.g. keep-alive).

**Comment by jmarantz** [comment]
Probably we should find a non-google approver to look at this after Kevin takes a look.

File: test/common/grpc/async_client_manager_impl_test.cc [hunk 1]
```diff
@@ -21,6 +24,80 @@ namespace Envoy {
 namespace Grpc {
 namespace {
 
+class RawAsyncClientCacheTest : public testing::Test {
+public:
+  RawAsyncClientCacheTest()
+      : api_(Api::createApiForTest(time_system_)),
+        dispatcher_(api_->allocateDispatcher("test_thread")), client_cache_(*dispatcher_) {}
+
+  void waitForSeconds(size_t seconds) {
```
Comment (jmarantz) [inline test/common/grpc/async_client_manager_impl_test.cc:33]: nit I'd use `int` here rather than size_t per convention/style guide.

File: test/common/grpc/async_client_manager_impl_test.cc [hunk 2]
```diff
@@ -21,6 +24,64 @@ namespace Envoy {
...
+    for (size_t i = 0; i < seconds; i++) {
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply test/common/grpc/async_client_manager_impl_test.cc:34]: still looking for the code comment here.

**Review by jmarantz — COMMENTED** [review]
looks great; just a few minor comments remain.

**Review by KBaichoo — DISMISSED** [review]

File: test/common/grpc/async_client_manager_impl_test.cc [hunk 3]
```diff
@@ -30,9 +30,11 @@ class RawAsyncClientCacheTest : public testing::Test {
...
         dispatcher_(api_->allocateDispatcher("test_thread")), client_cache_(*dispatcher_) {}
 
-  // Because advanceTimeAndRun simply increase the time system timestamp without interuption and
-  // then run the event loop, advance time in small steps so that the cache eviction timer can fire
-  // in correct timestamps.
+  // Because advanceTimeAndRun moves the current time as requested, and then executes
+  // all runnable timers in a non-deterministic order. This mimics real-time behavior in
+  // libevent if there is a long delay between libevent regaining control. Here we want to
+  // test behavior with a specific sequence of events, where each timer fires within a
+  // simulated second of what was programmed.
```
Comment (jmarantz) [inline test/common/grpc/async_client_manager_impl_test.cc:37]: sorry I did a minor edit after my initial 'send':
```
// advanceTimeAndRun moves the current time as requested, and then executes
// all runnable timers in a non-deterministic order. This mimics real-time behavior in
// libevent if there is a long delay between libevent regaining control. Here we want to
// test behavior with a specific sequence of events, where each timer fires within a
// simulated second of what was programmed.
```

**Review by KBaichoo — APPROVED** [review]

**Comment by jmarantz** [comment]
Thanks for pushing this through @chaoqin-li1123 . I think Snow would be a good non-googler senior maintainer to do the last round.

File: source/common/grpc/async_client_manager_impl.cc [hunk 4]
```diff
@@ -148,18 +149,58 @@ RawAsyncClientSharedPtr AsyncClientManagerImpl::getOrCreateRawAsyncClient(
...
+  if (lru_list_.size() == 1) {
+    evictEntriesAndResetEvictionTimer();
+  }
+}
+
+RawAsyncClientSharedPtr AsyncClientManagerImpl::RawAsyncClientCache::getCache(
+    const envoy::config::core::v3::GrpcService& config) {
+  RawAsyncClientSharedPtr client;
+  auto it = lru_map_.find(config);
+  if (it != lru_map_.end()) {
```
Comment (snowp) [inline source/common/grpc/async_client_manager_impl.cc:173]: nit: early return here (`return nullptr`) instead of putting all the logic inside the if branch

File: source/common/grpc/async_client_manager_impl.cc [hunk 5]
```diff
@@ -148,18 +149,58 @@ RawAsyncClientSharedPtr AsyncClientManagerImpl::getOrCreateRawAsyncClient(
...
+    auto cache_entry = it->second;
+    // Reset the eviction timer if the next entry to expire is accessed.
+    bool should_reset_timer = (cache_entry == --lru_list_.end());
```
Comment (snowp) [inline source/common/grpc/async_client_manager_impl.cc:176]: nit: const

File: source/common/grpc/async_client_manager_impl.cc [hunk 6]
```diff
@@ -148,18 +149,58 @@ RawAsyncClientSharedPtr AsyncClientManagerImpl::getOrCreateRawAsyncClient(
...
+    cache_entry->accessed_time_ = dispatcher_.timeSource().monotonicTime();
+    // Move the cache entry to the beginning of the list upon access.
+    lru_list_.splice(lru_list_.begin(), lru_list_, cache_entry);
+    // Get the cached async client before any cache eviction.
```
Comment (snowp) [inline source/common/grpc/async_client_manager_impl.cc:180]: Is the sequencing here important? I would think no since this entry is now the most recently used?

File: source/common/grpc/async_client_manager_impl.cc [hunk 7]
```diff
@@ -148,18 +149,58 @@ RawAsyncClientSharedPtr AsyncClientManagerImpl::getOrCreateRawAsyncClient(
...
+void AsyncClientManagerImpl::RawAsyncClientCache::evictEntriesAndResetEvictionTimer() {
+  MonotonicTime now = dispatcher_.timeSource().monotonicTime();
+  while (!lru_list_.empty()) {
+    MonotonicTime next_expire = lru_list_.back().accessed_time_ + EntryTimeoutInterval;
+    if (now >= next_expire) {
+      // Erase the expired entry.
+      lru_map_.erase(lru_list_.back().config_);
+      lru_list_.pop_back();
+    } else {
+      // Evict in a while loop because now timestamp keep changing, make sure next_expire > now
```
Comment (snowp) [inline source/common/grpc/async_client_manager_impl.cc:198]: This comment seems a bit out of place? I also don't see how `now` would keep changing since we only compute it once?

File: test/common/grpc/async_client_manager_impl_test.cc [hunk 8]
```diff
@@ -21,6 +24,85 @@ namespace Envoy {
...
+  // advanceTimeAndRun moves the current time as requested, and then executes
+  // all executable timers in a non-deterministic order. This mimics real-time behavior in
+  // libevent if there is a long delay between libevent regaining control. Here we want to
+  // test behavior with a specific sequence of events, where each timer fires within a
+  // simulated second of what was programmed.
+  void waitForSeconds(int seconds) {
+    for (int i = 0; i < seconds; i++) {
+      time_system_.advanceTimeAndRun(std::chrono::seconds(1), *dispatcher_,
+                                     Event::Dispatcher::RunType::NonBlock);
+    }
```
Comment (snowp) [inline test/common/grpc/async_client_manager_impl_test.cc:42]: Why do this in a for loop vs just advanceTimeAndRun for the full time?

File: test/common/grpc/async_client_manager_impl_test.cc [hunk 9]
```diff
@@ -21,6 +24,85 @@ namespace Envoy {
...
+TEST_F(RawAsyncClientCacheTest, CacheEviction) {
+  envoy::config::core::v3::GrpcService foo_service;
+  foo_service.mutable_envoy_grpc()->set_cluster_name("foo");
+  RawAsyncClientSharedPtr foo_client = std::make_shared<MockAsyncClient>();
+  client_cache_.setCache(foo_service, foo_client);
+  waitForSeconds(49);
+  // Cache entry hasn't been evicted because it was created 45s ago.
+  EXPECT_EQ(client_cache_.getCache(foo_service).get(), foo_client.get());
+  waitForSeconds(49);
+  // Cache entry hasn't been evicted because it was accessed 45s ago.
```
Comment (snowp) [inline test/common/grpc/async_client_manager_impl_test.cc:61]: Should this be 49s? Here and above

File: test/common/grpc/async_client_manager_impl_test.cc [hunk 10]
```diff
@@ -21,6 +24,85 @@ namespace Envoy {
...
+  }
+}
+
+TEST_F(RawAsyncClientCacheTest, GetTimeoutButNotEvictedCacheEntry) {
+  envoy::config::core::v3::GrpcService foo_service;
+  foo_service.mutable_envoy_grpc()->set_cluster_name("foo");
+  RawAsyncClientSharedPtr foo_client = std::make_shared<MockAsyncClient>();
+  client_cache_.setCache(foo_service, foo_client);
+  time_system_.advanceTimeAsyncImpl(std::chrono::seconds(50));
+  // Cache entry doesn't get evicted because it is accessed before timer fire.
```
Comment (snowp) [inline test/common/grpc/async_client_manager_impl_test.cc:98]: nit: I think the `hasn't been` phrasing used in the other test reads a bit better

File: test/common/grpc/async_client_manager_impl_test.cc [hunk 11]
```diff
@@ -21,6 +24,85 @@ namespace Envoy {
...
+    EXPECT_EQ(client_cache_.getCache(grpc_service).get(), nullptr);
+  }
+  // Cache entries 30s before haven't expired.
+  for (int i = 51; i <= 100; i++) {
+    grpc_service.mutable_envoy_grpc()->set_cluster_name(std::to_string(i));
+    EXPECT_EQ(client_cache_.getCache(grpc_service).get(), foo_client.get());
+  }
+}
+
+TEST_F(RawAsyncClientCacheTest, GetTimeoutButNotEvictedCacheEntry) {
```
Comment (snowp) [inline test/common/grpc/async_client_manager_impl_test.cc:92]: Can you add a comment to this test explaining what is being tested? The test name is a bit terse and hard to parse

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks, this seems pretty good to me, just a few nits and minor suggestions

File: source/common/grpc/async_client_manager_impl.cc (hunk 6 above)
Comment (chaoqin-li1123) [reply source/common/grpc/async_client_manager_impl.cc:180]: It is theoretically possible that the most recently used entry get evicted, imagine a very slow machine that take 50 s to execute a command and already expire the most recently entry when it start the eviction method(though very unlikely)

File: test/common/grpc/async_client_manager_impl_test.cc (hunk 8 above)
Comment (chaoqin-li1123) [reply test/common/grpc/async_client_manager_impl_test.cc:42]: Please take a look at the method comment, because advanceTimeAndRun won't wake up timer in the middle of the function. but a real time system will invoke timer as time advance.

Comment (snowp) [reply test/common/grpc/async_client_manager_impl_test.cc:42]: Ok thanks, I missed the comment

**Review by htuch — APPROVED** [review]
